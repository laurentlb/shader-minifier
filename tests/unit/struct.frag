#version 460
// struct declaration in TypeDecl
struct MarchData { // newName
    float dist;
    vec3 col;
    float spec;
};
// struct in TLDecl
MarchData m;
uniform MyInterfaceBlock
{
    // struct in interface block field
    MarchData march;
    /* Not allowed, at least on shadertoy.
    // anonymous struct declaration in interface block field
    struct {
        // struct identifier use in field in interface block
        MarchData abc;
    } struc;
    */
};
// struct identifier use in FunctionType.retType
MarchData struct_as_ret()
{
    return m;
}
// struct identifier use in FunctionType.args
int struct_as_arg(MarchData m)
{
    return 1;
}
int struct_uses() {
    // struct identifier use in Stmt.Decl
    MarchData m;
    // struct field access
    m.dist = 1.;
    m.spec = m.dist / 2.;
    // struct identifier use in constructor
    m = MarchData(1., vec3(1), m.spec);
    // anonymous struct declaration in Stmt.Decl
    struct {
        // struct identifier use in field in local
        MarchData abc;
    } s;
    // struct identifier use in Stmt.Decl
    for (MarchData m2; true; ) break;
    // anonymous struct declaration in Stmt.ForD
    for (struct {
                    // struct identifier use in field in local
                    MarchData abc;
                } s; true; ) break;
    return int(m.col.x + s.abc.dist + march.spec);
}

// anonymous struct declaration at top level
struct {
    // struct identifier use in field in global struct
    MarchData abc;
} s2;

// bug test
float v(vec4 w)
{
  float MarchData = w.x-4.0; // reuse type name as var
  return w.w + MarchData*MarchData; // don't rename field `.w`
}

struct Foo {int A; int B;};
struct Bar {int A; int C; int B;};

Foo foo;
Bar bar;

void main() {
    foo.A = 1;
    foo.B = 2;
    bar.A = 3;
    bar.C = 4;
    bar.B = 5;
}

struct Baz {int A; int B;} baz;
struct {int A; int B;}; // why is this even legal
