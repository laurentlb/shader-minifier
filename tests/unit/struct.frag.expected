#version 460

struct r{float e;vec3 k;float M;};
r i;
uniform MyInterfaceBlock{r B;};
r t()
{
  return i;
}
int t(r i)
{
  return 1;
}
int I()
{
  r i;
  i.e=1.;
  i.M=i.e/2.;
  i=r(1.,vec3(1),i.M);
  struct{r O;} P;
  for(r i;true;)
    break;
  for(struct{r I;} i;true;)
    break;
  return int(i.k.x+P.O.e+B.M);
}
struct{r P;} O;
float I(vec4 i)
{
  float P=i.x-4.;
  return i.w+P*P;
}
